docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+a "request unit". There will be a direct conversion from RUs to a dollar amount
+that is part of a tenant's bill. Request units are determined based on CPU usage
+in the SQL pods and on the KV requests from the tenant. For more details, see
+the "Managing Serverless Resource Usage" RFC
+([internal-only](https://github.com/cockroachlabs/managed-service/blob/master/docs/RFCS/20210603_serverless_usage.md)).
+
+Each tenant will have a number of RUs available to use as needed, plus a
+"baseline" rate of replenishment. These map directly to the concept of a
+[token bucket](https://en.wikipedia.org/wiki/Token_bucket).
+The values are determined by a higher-level billing algorithm that is beyond the
+scope of this RFC and need to be configurable on-the-fly.
+
+The enforcement of RU consumption will happen inside the SQL pods, at the point
+where KV requests are sent. The SQL pods communicate with centralized state (for
+that tenant) which lives in a system table inside the system tenant. The
+centralized state also tracks and reports the consumption of RUs.
+
+#### Impact
+
+The new subsystem will only be enabled on our multi-tenant clusters. The
+subsystem involves overhead related to communicating with and updating the
+centralized state for each tenant.
+
+Note that admission control for preventing overload in the host cluster or for
+reducing the variability in user-visible performance (depending on other load on
+the host cluster) are not in the scope of this RFC.
+
+# Motivation
+
+The higher-level motivation is covered by the "Managing Serverless Resource
+Usage" RFC ([internal-only](https://github.com/cockroachlabs/managed-service/blob/master/docs/RFCS/20210603_serverless_usage.md)).
+This RFC covers the implementation details of a subsystem necessary to enforce
+resource usage.
+
+# Technical design
+
+## Overview
+
+We implement a "global" (distributed) token bucket for each tenant; the metadata
+for all these buckets lives in a system table on the system tenant. The global
+bucket is configured with:
+ - an initial amount of "burst" tokens (RUs)
+ - a refill rate (RUs/second)
+ - a limit on how many unused tokens (RUs) we can accumulate as burst. Refill is
+   essentially paused when the bucket has more tokens than the limit.
+
+Each SQL pod implements a local token bucket and uses it for admission control
+at the KV client level (i.e. before sending out KV requests to the host
+cluster). Each KV request consumes tokens and might need to wait until there are
+enough tokens available.  The local token bucket is also adjusted periodically
+to account for CPU usage on the pod.
+
+The local bucket is periodically filled with tokens from the global bucket, or
+assigned a fraction of the global fill rate. This happens via API calls through
+the KV connector (see `kvtenant.Connector`). Each call is implemented as a
+transaction involving the system table.
+
+The global bucket keeps track of cumulative RU usage, which is reported
```

Comment:
Whenever a KV node services a request to the distributed token bucket, it updates an in-memory per-tenant stat. The stats would be scraped from all KV nodes, and the latest (or greatest total consumption) datapoint for each tenant would be used. AFAICT this is almost "free" and avoids the complexity with dealing with scraping the tenant side.

One could make the argument that we would want to scrape stats from the tenant side too just to get more up-to-date stats (especially if we want to increase the bucket update interval), but I'm not sure the benefit is worth the complexity (it would require some kind of reconciliation when a node crashes).

Reply:
But then why are the stats stored on each write via the `  -- Cumulative usage statistics.` fields? Updating them in-mem and scraping them from there is fine, but it does mean that you need to do the "integrate over the derivative" thing on the data (summed over all host nodes) to get the total consumption. (Which, might be fine, it's definitely the most idiomatic use of metrics here, but again I'm missing something because nothing there requires putting the stats into the table).

Reply:
I thought it's a relatively cheap way to have a single "source of truth" for this information, which will factor into the customer's bill. We are writing a new row anyway, do you think there is a problem with including a few more values in the row?

The in-memory stat will contain the current cumulative "since-beginning-of-time" values, so you don't need to do any integration. You just need the latest values (and the values at some earlier time, e.g. billing cycle start)

Reply:
So is the actual suggestion the following:

- each node has these metrics, but initially they're zero (or rather, have nothing for any tenant)
- when a node first serves a quota request for a tenant X, "on the way out" it populates the metric gauges under tenant X's labels from the write that just happened to that tenant's table (which that node is aware of)
- billing looks at `max(tenant=X, nodes=any)` to get the absolute consumption

I like that, sorry, it took me a long time to understand that, please add a dumbed down version of that to the RFC.

> We are writing a new row anyway, do you think there is a problem with including a few more values in the row?

I suppose not, if we're ok with writing all of these rows we're probably ok with ~duplicating (?) their size.